conformalInference/R/multisplit.R

```diff
@@ -0,0 +1,193 @@
+#'   values, instead of calling the conformal function separately m = 5 times.
+#'
+#' @details This function is based on the package \code{\link{future.apply}} to
+#'  perform parallelization. If this package is not installed, then the function
+#'   will proceed serially.
+#'
+#' @seealso \code{\link{conformal.pred.split}}
+#' @references "Multi Split Conformal Prediction" by Solari, Djordjilovic (2021).
+#' @export conformal.pred.msplit
+
+
+
+conformal.pred.msplit = function(x, y, x0,train.fun, predict.fun, alpha=0.1,
+                                 rho = rep(0.5,B), w=NULL, mad.train.fun=NULL, mad.predict.fun=NULL,
+                                 split=NULL, seed=NULL,
+                                 verbose=FALSE,
+                                 B=50,
+                                 lambda = 0.5,
+                                 tau = 1-(B+1)/(2*B)) {
+
+
+  x = as.matrix(x)
+  y = as.numeric(y)
+  n = nrow(x)
+  p = ncol(x)
+  x0 = matrix(x0,ncol=p)
+  n0 = nrow(x0)
+  
+  ## Is the future.apply library installed?
+  paral = "future.apply" %in% rownames(installed.packages())
+  
+
+  check.args(x=x,y=y,x0=x0,alpha=alpha,train.fun=train.fun,
+             predict.fun=predict.fun,mad.train.fun=mad.train.fun,
+             mad.predict.fun=mad.predict.fun)
+
+
+  if (!is.null(seed)) 
+    set.seed(seed)
+  else{
+    seed=B
+  }
+  
+  
+  if (is.null(rho) || !is.vector(rho) || length(rho)!=B){
+    #print("rho value either not provided or with uncorrect format. 0.5 is set as
+    #      default value in all the B replications!")
+    rho = rep(0.5,B)
+  }
+  
+  split.fit = is.matrix(split) && (nrow(split) == B) && (ncol(split) > 0 && ncol(split) < n)
+  
+  if(!split.fit && !is.null(split)){
+    stop("The 'split' argument should either be NULL or a matrix of dimension B x n1,
+         where n1 is the number of observation in the training set !  ")
+  }
+  
+
+  loB <- upB <- matrix(NA,nrow=n0,ncol=B)
+  tr <- tau*B + .001
+  
+  check.pos.num(lambda)
+  check.num.01(tau)
+
+  if(paral){
+    #Define structures for parallelization
+    future::plan(future::multisession)
+    options(future.rng.onMisuse="ignore")
+}
+
+  ## Compute the B prediction intervals
+
+    lo_up<- one.lapply(1:B, function(bbb) {
+      
+      prop=rho[bbb]
+
+      out<-conformal.pred.split(x=x,y= y, x0=x0, train.fun=train.fun, 
+                                predict.fun=predict.fun,
+                                alpha=alpha*(1-tau) + (alpha*lambda)/B, 
+                                split=split[bbb,],rho = prop,
+                                w=w, mad.train.fun=mad.train.fun,
+                                mad.predict.fun=mad.predict.fun,
+                                seed=seed+bbb,
+                                            verbose=verbose)
+
+
+      return(cbind(t(out$lo),t(out$up)))
+    })
+
+
+  # Redefine the appropriate data structures
+
+  Y_lo_up=do.call(rbind, lo_up)
+  m=nrow(lo_up[[1]])
+  index=rep(rep(1:m),B)
```

Comment:
It would be simpler to include this index as a column in line 154 you return the prediction intervals. For example, you could just change that line to `return(cbind(t(out$lo), t(out$up), 1:m))`.